optuna/integration/mlflow.py

```diff
@@ -239,14 +247,37 @@ def _set_tags(self, trial: optuna.trial.FrozenTrial, study: optuna.study.Study)
         # This sets the tags for MLflow.
         mlflow.set_tags(tags)
 
-    def _log_metric(self, value: Union[float, None]) -> None:
-        """Log the trial result as metric to MLflow.
+    def _log_metrics(self, values: List[Union[float, None]]) -> None:
+        """Log the trial results as metrics to MLflow.
 
         Args:
-            value: Result of trial.
+            values: Results of a trial.
         """
-        trial_value = value if value is not None else float("nan")
-        mlflow.log_metric(self._metric_name, trial_value)
+
+        if not isinstance(self._metric_name, str):
+            if len(self._metric_name) != len(values):
+                raise ValueError(
+                    "Running multi-objective optimization "
+                    "with {} objective values, but {} names specified. "
+                    "Match objective values and names, or use default broadcasting.".format(
+                        len(values), len(self._metric_name)
+                    )
+                )
+
+            else:
+                names = self._metric_name
+
+        else:
+            if len(values) > 1:
+                # Broadcast default name for multi-objective optimization.
+                names = ["{}_{}".format(self._metric_name, i) for i in range(len(values))]
+
+            else:
+                names = [self._metric_name]
```

Comment:
Thank you for relaxing the type of `metric_name`. 
I feel this `not isinstance(self._metric_name, str)` is a bit ambiguous. Non-string objects will be processed in the if clause, so a dictionary, for example, will also processed there. It raises the following error, but I guess we can show more user-friendly error message.

```
TypeError: expected string or bytes-like object
```

The following is just a suggestion of changes.

```suggestion
        if isinstance(self._metric_name, str):
            if len(values) > 1:
                # Broadcast default name for multi-objective optimization.
                names = ["{}_{}".format(self._metric_name, i) for i in range(len(values))]
            else:
                names = [self._metric_name]
        else:
            if not isinstance(self._metric_name, Sequence):
                raise TypeError(
                    "metric_name is supposed to be a string or a sequence of strings."
                )
            if len(self._metric_name) != len(values):
                raise ValueError(
                    "Running multi-objective optimization "
                    "with {} objective values, but {} names specified. "
                    "Match objective values and names, or use default broadcasting.".format(
                        len(values), len(self._metric_name)
                    )
                )
            else:
                names = self._metric_name
```

In addition, how about adding a test case corresponding to the type checking?

```python
def test_multiobjective_raises_on_type_mismatch(tmpdir: py.path.local) -> None:

    tracking_file_name = "file:{}".format(tmpdir)
    mlflc = MLflowCallback(tracking_uri=tracking_file_name, metric_name={0: "foo", 1: "bar"})
    study = optuna.create_study(study_name="my_study", directions=["minimize", "maximize"])

    with pytest.raises(TypeError):
        study.optimize(_multiobjective_func, n_trials=1, callbacks=[mlflc])
```

Reply:
How do you feel about checking for `Sequence` at the beginning, before hitting this `if else` clause? Something like

```python
if not isinstance(self._metric_name, Sequence):
    raise ValueError("metric_name is supposed to be a string or a sequence of strings.")


if not isinstance(self._metric_name, str):
[...]
```